stamina-testkit/src/main/scala/stamina/testkit/StaminaTestKit.scala

```diff
@@ -0,0 +1,85 @@
+package stamina
+package testkit
+
+import scala.util._
+import scala.annotation.tailrec
+
+trait StaminaTestKit { self: org.scalatest.WordSpecLike ⇒
+
+  val defaultSampleId = "default"
+  case class PersistableSample(sampleId: String, persistable: AnyRef, description: Option[String]) {
+    override def toString = persistable.getClass.getSimpleName + description.map(" " + _).getOrElse("")
+  }
+
+  def sample(persistable: AnyRef) = new PersistableSample(defaultSampleId, persistable, None)
+  def sample(sampleId: String, persistable: AnyRef) = new PersistableSample(sampleId, persistable, Some(sampleId))
+  def sample(sampleId: String, persistable: AnyRef, description: String) = new PersistableSample(sampleId, persistable, Some(description))
+
+  implicit class TestablePersisters(persisters: Persisters) extends org.scalatest.Matchers {
+    def generateTestsFor(samples: PersistableSample*): Unit = {
+      samples.foreach { sample ⇒
+        performRoundtrip(sample)
+        deserializeStoredVersions(sample)
+      }
+      // Here we could verify test coverage
+    }
+
+    private def performRoundtrip(sample: PersistableSample) = {
+      s"persist and unpersist $sample" in {
+        persisters.canPersist(sample.persistable) should be(true)
+        persisters.unpersist(persisters.persist(sample.persistable)) should equal(sample.persistable)
+      }
+    }
+
+    @tailrec
+    private def deserializeStoredVersions(sample: PersistableSample, fromVersion: Int = 1): Unit = {
+
+      s"deserialize the stored serialized form of $sample version $fromVersion" in {
+        verifyByteStringDeserialization(sample, fromVersion)
+      }
+
+      if (fromVersion < latestVersion(sample.persistable))
+        deserializeStoredVersions(sample, fromVersion + 1)
+    }
```

Comment:
This might simply be a matter of style or personal taste but why not drop the tail recursion and just do something like:

```scala
Range(fromVersion, latestVersion(sample.persistable)).foreach { version =>
  s"deserialize the stored serialized form of $sample version $version" in {
    verifyByteStringDeserialization(sample, fromVersion)
  }
}
```

or, if you prefer the alternative style:

fromVersion until latestVersion(sample.persistable)) foreach { version =>
  s"deserialize the stored serialized form of $sample version $version" in {
    verifyByteStringDeserialization(sample, fromVersion)
  }
}
```